api/agent/agent.go

```diff
@@ -113,13 +113,39 @@ type agent struct {
 	// TODO(reed): shoot this fucking thing
 	callOverrider CallOverrider
 
+	// Call Extension interceptor
+	extensionInterceptor ExtensionInterceptor
+
 	// additional options to configure each call
 	callOpts []CallOpt
 
 	// deferred actions to call at end of initialisation
 	onStartup []func()
 }
 
+type ExtensionInterceptor interface {
+	// BeforeCreate is triggered immediately preceding container creation, after all other
+	// configuration is in place.
+	// ContainerTask.Id() can be used to provide a stable per-container key.
+	BeforeCreate(context.Context, drivers.ContainerTask, drivers.Cookie, Call) error
+
+	// AfterDestroy is triggered immediately after container shutdown.
+	// ContainerTask.Id() can be used to provide a stable per-container key.
+	AfterDestroy(context.Context, drivers.ContainerTask) error
+
+	// BeforeCall is triggered immediately preceding call invocation.
+	// ContainerTask.Id() can be used to provide a stable per-container key that corresponds
+	// to the Id seen by BeforeCreate/AfterDestroy.
+	BeforeCall(context.Context, drivers.ContainerTask, Call, CallExtensions) error
+
+	// AfterCall is triggered immediately after an invocation, to tidy up any per-invocation state.
+	// ContainerTask.Id() can be used to provide a stable per-container key that corresponds
+	// to the Id seen by BeforeCreate/AfterDestroy.
+	AfterCall(context.Context, drivers.ContainerTask, Call, CallExtensions) error
+}
```

Comment:
the rest of these kinds of things are defined in the `fnext` package, could we move these there as well or is there a reason to keep them in agent? they are basically stubs, so they are easy to move around I think. to fit in with the other kids at school, perhaps `ContainerListener` would be more fitting than `ExtensionInterceptor` after moving to that package, with some more filled in method names to match, too, most likely, e.g. `BeforeContainerCreate`, etc.? feel free to take more artistic license with the names..

Reply:
Sure.